### PR TITLE
Ramp rates

### LINKED ISSUE #88: [TASK] set a ramp-rate constraint
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
I hope HERON has a built-in feature that implements ramp-rate constraints on generators. 

**Describe the solution you'd like**

**Describe alternatives you've considered**

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
##### What issue does this change request address?
Closes #88, and contains all of PR #284 

##### What are the significant changes in functionality due to this change request?
Adds ramp rate limiters to `produces` components (not `demands` or `stores`).

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/HERON/wiki/Code-Standards) for details.
- [x] 4. Automated Tests should pass.
- [x] 5. If significant functionality is added, there must be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large tes.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added, the the analytic documentation must be updated/added.
- [x] 9. If any test used as a basis for documentation examples have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by PaulTalbot-INL** [comment]
Screenshot 2023-04-19 at 15 49 40
I did not see this locally before, looking at it more carefully ...

**Comment by PaulTalbot-INL** [comment]
hm, not sure why this is failing and not getting enough info from the tester to see what it's failing on. I'll have to try and find a windows machine to test on.

**Review by dylanjm — DISMISSED** [review]
Changes LGTM

**Comment by PaulTalbot-INL** [comment]
Well, @dylanjm, relative tolerance didn't work before, but seems to work in conjunction with ordered CSV. Good enough for me if it's good enough for you.

File: templates/template_driver.py [hunk 1]
```diff
@@ -1100,25 +1145,25 @@ def _modify_inner_optimization_settings(self, template, case):
       # add optimization objective to PostProcessor list if not already there
       pp_node = template.find('Models').find(".//PostProcessor[@name='statistics']")
       if new_objective != 'missing':
-        raven_metric_name = case._optimization_settings['metric']['name']
-        prefix = case.metrics_mapping[raven_metric_name]['prefix']
+        raven_metric_name = optimization_settings['stats_metric']['name']
+        prefix = case.stats_metrics_mapping[raven_metric_name]['prefix']
```
Comment (dylanjm) [inline templates/template_driver.py]: I'm worried some of these changes are accidently being imported from Gabe's previous PR. Did you mean to make these changes?

File: src/DispatchPlot.py [hunk 2]
```diff
@@ -115,38 +117,69 @@ def plot_component(self, fig, axes, df, grp_vars, comp_idx, sid, mstep, cid) ->
...
+      @ In, cdict, Dict[str, str], a dictionary contains color code to variables
       @ Out, None
     """
+    # Pre-define color codes and transparency
+    Gray, Dark = ('#dcddde','#1a2b3c')
+    alpha = '70'
     for (key, group), ax in zip(grp_vars.items(), axes.flat):
-      lines = []
+      # Define list for data, label, and color. Seperate 'level'(line plot) with other variables (stack plot)
+      positive_dat = []
```
Comment (dylanjm) [inline src/DispatchPlot.py]: I'm worried some of these changes are being imported from Xingyue's previous PRs. Did you mean to include these changes?

**Review by dylanjm — APPROVED** [review]
LGTM

**Pull request merged**